### PR TITLE
Migrate pausable to cairo2

### LINKED ISSUE #655: Migrate Pausable to Cairo 2

### PR DESCRIPTION
**Pull request by andrew-fleming**
Fixes #655 

Depends on #661

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

### CODE REVIEW

File: src/openzeppelin/security/pausable.cairo [hunk 1]
```diff
@@ -1,39 +1,60 @@
...
+        }
+
+        fn assert_paused(self: @ContractState) {
+            assert(self.paused.read(), 'Pausable: not paused');
+        }
+
+        fn _pause(ref self: ContractState) {
+            self.assert_not_paused();
+            self.paused.write(true);
+            self.emit(Event::Paused(Paused { account: get_caller_address() }));
```
Comment (ericnordelo) [inline src/openzeppelin/security/pausable.cairo:51]:
```suggestion
            self.emit(Paused { account: get_caller_address() });
```

I think this should work as well, right?

File: src/openzeppelin/security/pausable.cairo [hunk 2]
```diff
@@ -1,39 +1,60 @@
...
+        }
+
+        fn _unpause(ref self: ContractState) {
+            self.assert_paused();
+            self.paused.write(false);
+            self.emit(Event::Unpaused(Unpaused { account: get_caller_address() }));
```
Comment (ericnordelo) [inline src/openzeppelin/security/pausable.cairo:57]:
```suggestion
            self.emit(Unpaused { account: get_caller_address() });
```

File: src/openzeppelin/tests/security/test_pausable.cairo [hunk 3]
```diff
@@ -1,47 +1,107 @@
...
 #[available_gas(2000000)]
-fn test_pause_when_unpaused() {
-    assert(!MockPausable::is_paused(), 'Should not be paused');
-    assert(MockPausable::get_count() == 0, 'Should be 0');
-    MockPausable::assert_unpaused_and_increment();
-    assert(MockPausable::get_count() == 1, 'Should increment');
-    MockPausable::pause();
-    assert(MockPausable::is_paused(), 'Should be paused');
+fn test_is_paused() {
+    let mut state = STATE();
```
Comment (ericnordelo) [inline src/openzeppelin/tests/security/test_pausable.cairo:16]:
```suggestion
    let state = STATE();
```

File: src/openzeppelin/tests/security/test_pausable.cairo [hunk 4]
```diff
@@ -1,47 +1,107 @@
...
+    InternalImpl::_pause(ref state);
+    InternalImpl::assert_paused(@state);
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected: ('Pausable: not paused', ))]
+fn test_assert_paused_when_not_paused() {
+    let mut state = STATE();
```
Comment (ericnordelo) [inline src/openzeppelin/tests/security/test_pausable.cairo:42]:
```suggestion
    let state = STATE();
```

File: src/openzeppelin/tests/security/test_pausable.cairo [hunk 5]
```diff
@@ -1,47 +1,107 @@
...
+fn test_assert_not_paused_when_paused() {
+    let mut state = STATE();
+    InternalImpl::_pause(ref state);
+    InternalImpl::assert_not_paused(@state);
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_assert_not_paused_when_not_paused() {
+    let mut state = STATE();
```
Comment (ericnordelo) [inline src/openzeppelin/tests/security/test_pausable.cairo:62]:
```suggestion
    let state = STATE();
```

**Review by ericnordelo — COMMENTED** [review]
Looking good Andrew! Left some small suggestions, but besides that I think is good to go.

File: src/openzeppelin/security/pausable.cairo (hunk 1 above)
Comment (andrew-fleming) [reply src/openzeppelin/security/pausable.cairo:51]: Good call! This is much easier

File: src/openzeppelin/tests/security/test_pausable.cairo (hunk 3 above)
Comment (andrew-fleming) [reply src/openzeppelin/tests/security/test_pausable.cairo:16]: We need this to be mutable to call `_pause` and `_unpause` in this test

**Review by ericnordelo — APPROVED** [review]
LGTM!

**Pull request merged**